ouroboros-network/src/Ouroboros/Network/Mux/Control.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies   #-}
+
+module Ouroboros.Network.Mux.Control where
+
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Codec.CBOR.Term as CBOR
+import           Codec.CBOR.Write (toLazyByteString)
+import           Codec.Serialise.Class
+import           Control.Exception
+import           Control.Monad
+
+import qualified Data.Map.Strict as M
+import           Data.Maybe (mapMaybe)
+import           GHC.Stack
+
+import           Ouroboros.Network.Mux.Types
+
+data ControlMsg = MsgInitReq [Version]
+                | MsgInitRsp Version
+                | MsgInitFail String
+                deriving Show
+
+encodeCtrlMsg :: ControlMsg -> CBOR.Encoding
+encodeCtrlMsg (MsgInitReq versions) = CBOR.encodeListLen 2 <> CBOR.encodeWord 0 <> encodeVersionMap versions
+encodeCtrlMsg (MsgInitRsp version)  = CBOR.encodeListLen 2 <> CBOR.encodeWord 1 <> encode version
+encodeCtrlMsg (MsgInitFail msg)     = CBOR.encodeListLen 2 <> CBOR.encodeWord 2 <> encode msg
+
+encodeVersionMap :: [Version] -> CBOR.Encoding
+encodeVersionMap versions =
+    CBOR.encodeMapLen (fromIntegral $ length versions)
```

Comment:
Hmm, why encode the list as a map and not as a list?

Reply:
In CBOR a map is a list of pairs of elements which is what we have. A version number and a version `Term`.